Fixes #27264 - Install ansible-runner package

ezr-ondrej opened the pull request:
As ansible-runner is going to be default way to run ansible, it should become a dependency at install time.

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,38 @@
+    'RedHat': {
+      yumrepo { 'ansible-runner':
+        descr    => "Ansible runner",
+        baseurl  => "baseurl=https://releases.ansible.com/ansible-runner/rpm/epel-7-x86_64/",
```
```suggestion
        baseurl  => "https://releases.ansible.com/ansible-runner/rpm/epel-${facts['os']['release']['major']}-x86_64/",
```

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,38 @@
+      ::apt::source { 'ansible-runner':
+        repos    => 'main',
+        location => 'https://releases.ansible.com/ansible-runner/deb',
+        key      => 'AC48AC71DA695CA15F2D39C4B84E339C442667A9',
```
You can set all the key info here as well and drop the `apt::key` definition.
```suggestion
        key      => {
          id     => 'AC48AC71DA695CA15F2D39C4B84E339C442667A9',
          source => 'https://releases.ansible.com/keys/RPM-GPG-KEY-ansible-release.pub',
        },
```

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,38 @@
+      }
+    }
+    default: {
+      $package = 'python-ansible-runner'
```
This will fail because you can't change variables in Puppet. That means you need to define the package name in the various branches

ezr-ondrej commented:
Thanks @ekohl for the advices, I have updated the class to use external repositories conditionally, and added some tests.
Could you plese take another look? :)

ezr-ondrej commented on manifests/plugin/ansible/params.pp:
```diff
@@ -6,4 +6,5 @@
   $working_dir = '/tmp'
   $host_key_checking = false
   $stdout_callback = 'yaml'
+  $upstream_runner_repo = true
```
is this enough to get picked up by `kafo`, or do I need to do something else?

ezr-ondrej commented on manifests/plugin/ansible.pp:
```diff
@@ -29,6 +29,7 @@
   Optional[Stdlib::Absolutepath] $working_dir = $::foreman_proxy::plugin::ansible::params::working_dir,
   Boolean $host_key_checking = $::foreman_proxy::plugin::ansible::params::host_key_checking,
   String $stdout_callback = $::foreman_proxy::plugin::ansible::params::stdout_callback,
+  Boolean $upstream_runner_repo = $::foreman_proxy::plugin::ansible::params::upstream_runner_repo,
```
Is it necessary here?

ekohl replied:
This exposes it to kafo and users. A more conventional naming is `manage_runner_repo`

ezr-ondrej replied:
I've renamed the param

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+# $upstream_runner_repo:: If true, adds upstream repositories to install ansible-runner package from
+#
+class foreman_proxy::plugin::ansible::runner(
+  Boolean $upstream_runner_repo = $foreman_proxy::plugin::ansible::params::upstream_runner_repo,
```
I think you can directly refer to the ansible class as well. Since you're passing it in ansible.pp, you can also leave out the default.

```suggestion
  Boolean $upstream_runner_repo = $foreman_proxy::plugin::ansible::upstream_runner_repo,
```

ezr-ondrej replied:
I have found out, I do not need to pass the param out, if I am using it like you are suggesting. Is that correct way? Test are passing...

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+  Boolean $upstream_runner_repo = $foreman_proxy::plugin::ansible::params::upstream_runner_repo,
+) {
+
+  if ( $upstream_runner_repo ) {
```
You can leave out the parenthesis here

```suggestion
  if $upstream_runner_repo {
```

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+      'RedHat': {
+        yumrepo { 'ansible-runner':
+          descr    => 'Ansible runner',
+          baseurl  => 'https://releases.ansible.com/ansible-runner/rpm/epel-7-x86_64/',
```
To make sure it'll also work with EL8 you can use variables. I think there's also one for the architecture
```suggestion
          baseurl  => 'https://releases.ansible.com/ansible-runner/rpm/epel-${facts['os']['release']['major']}-x86_64/',
```

ezr-ondrej replied:
well.. not that `ansible-runner` would have builds for others - https://releases.ansible.com/ansible-runner/rpm/ xD but I guess it is for now and they will keep them comming.

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+        $package = 'ansible-runner'
+      }
+      default: {
+        $package = 'python-ansible-runner'
```
Seeing as how all of them are `ansible-runner`, is this for a BSD? If you don't know, I'd start with simply adding a class parameter `String $package_name = 'ansible-runner'` to keep it simple.

ezr-ondrej replied:
It was just some dark thoughts leading me that direction, I added warning for other OSs, you think it is enough?

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+) {
+
+  if $manage_runner_repo {
+    case $::osfamily {
```
Let's start using the modern facts (we still need to convert the rest of the module):
```suggestion
    case $facts['os']['family'] {
```

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,48 @@
+    case $::osfamily {
+      'Debian': {
+        include ::apt
+        ::apt::source { 'ansible-runner':
```
The leading colons are no longer needed. The rest of the module still has to be converted.

```suggestion
        apt::source { 'ansible-runner':
```

ezr-ondrej commented:
Thanks :+1: I followed your advice once again :blue_heart:

ezr-ondrej commented:
Something more to add, @ekohl? I would be happy to hear any more improvement suggestions :yum:

ares commented:
looks good to me, @ekohl could you please give it a quick look? this is 1.23 feature

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,47 @@
+        }
+      }
+      default: {
+        warning("Repository containing 'ansible-runner' not known for '${facts['os']['family']}'")
```
Since we have a parameter to turn this off, I think it's fine to hard fail on this:

```suggestion
        fail("Repository containing 'ansible-runner' not known for '${facts['os']['family']}'")
```

ezr-ondrej replied:
yeah, you are right :+1:

ekohl commented on spec/classes/foreman_proxy__plugin__ansible_spec.rb:
```diff
@@ -45,16 +68,25 @@
+            :working_dir         => '/tmp/ansible',
+            :host_key_checking   => true,
+            :stdout_callback     => 'debug',
+            :manage_runner_repo  => false
```
Nit: could you add the trailing comma? We should enforce this via rubocop but until then ... :)

While you're changing this, mind switching it to `manage_runner_repo: false` style and removing the alignment? I've been doing this in other places in our test code because the alignment just creates noise.

ezr-ondrej replied:
I like it much more actually :)

ezr-ondrej commented:
I've added the issue, sorry for not doing it.

ekohl commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  package { $package_name:
+    ensure => 'latest',
```
In the installer we always use installed or make it configurable. The reason is that this module is also used by people outside of the installer and it's always very painful if a new releases comes out that actually breaks your operation outside of a scheduled window. From personal experience I know how painful this can be.
```suggestion
    ensure => 'installed',
```

ezr-ondrej commented:
Do not apologize for making my code better! :stuck_out_tongue:
I appreciate your patience :blush: :+1:

ekohl commented:
Also looks like I just introduced merge conflicts :(

ares commented:
Just to be clear about the next steps
1) rebase
2) PR to upgrade notes under theforeman.org repo 
3) start a thread on discourse aiming not devs but users, probably with ansible tag, asking what should be the default and if the linked repo raises concerns

With all of above, this can be merged :+1:

tbrisker commented:
What's the status here? iiuc this is blocking the module release for 1.23?

ares commented:
yes, I think we're ready to merge now, /me looks at @ekohl, I suppose the discourse thread should start once this is in and upgrade notes are in place (that one is also blocked on this to be merged)

ekohl commented on manifests/plugin/ansible.pp:
```diff
@@ -54,6 +57,7 @@
   }
 
   include ::foreman_proxy::plugin::dynflow
+  include ::foreman_proxy::plugin::ansible::runner
```
Do we actually need a way to turn this off? If you don't want the external repo but don't have an alternative then it'll fail.

ezr-ondrej replied:
It will, but if you do not have alternative, than the ansible will fail as of 1.24, so I do not see a problem in failing, maybe some better warning?

ekohl replied:
If it's a hard requirement, packaging should require it. That will have some implications so would be good to have a heads up there so we tackle it in time.

ezr-ondrej replied:
Well it is hard requirement as of 1.24, so we can have a parameter disabling it for 1.23 and than remove it once it is hard requirement ?

ezr-ondrej commented:
I've added parameter `install_runner` to disable install `ansible-runner` all together. Is that what you've meant, @ekohl?

tbrisker commented:
This seems to be failing on debian

evgeni commented on manifests/plugin/ansible/runner.pp:
```diff
@@ -0,0 +1,47 @@
+      'RedHat': {
+        yumrepo { 'ansible-runner':
+          descr    => 'Ansible runner',
+          baseurl  => "https://releases.ansible.com/ansible-runner/rpm/epel-${facts['os']['release']['major']}-${facts['architecture']}/",
```
```suggestion
          baseurl  => "https://releases.ansible.com/ansible-runner/rpm/epel-${facts['os']['release']['major']}-\$basearch/",
```

Not needed as long as we only have x86_64 repos, but Puppets `arch` can be different from what rpm/yum/dnf define as `basearch`

ezr-ondrej replied:
Thanks :)

ekohl commented on spec/classes/foreman_proxy__plugin__ansible_spec.rb:
```diff
@@ -8,12 +8,35 @@
+        when 'redhat-7-x86_64'
+          it 'should include ansible-runner upstream repo' do
+            should contain_yumrepo('ansible-runner')
+                   .with_baseurl("https://releases.ansible.com/ansible-runner/rpm/epel-7-x86_64/")
```
```suggestion
                   .with_baseurl("https://releases.ansible.com/ansible-runner/rpm/epel-7-$basearch/")
```

ekohl commented on spec/classes/foreman_proxy__plugin__ansible_spec.rb:
```diff
@@ -82,6 +110,14 @@
+          { install_runner: false }
+        end
+
+        it { should_not contain_package('ansible-runner') }
```
Could you add the repo checks here as well (that they shouldn't be present)? You can also add `it { should_not contain_class('foreman_proxy::plugin::ansible::runner') }`

ezr-ondrej replied:
I did both :+1:

ekohl commented:
Made a change that should fix the tests. Will squash on merge when the unit tests are green.

tbrisker commented:
Looks like it still isn't working on debian:
```
  Error: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install foreman-proxy' returned 100: Reading package lists...
  Building dependency tree...
  Reading state information...
  E: Unable to locate package foreman-proxy

ekohl commented:
That's a false positive. See the timeout before it. This is also failing in master and this has no acceptance tests.

It's pulling in puppetlabs-apt 2.0.1 because another module is already installing puppetlabs-translate 2.0 and the latest ones are incompatible with that version. 2.0.1 is the last version that doesn't pull in translate. Now technically we should probably mark incompatible with such an old version if it's broken, but that would block the entire acceptance test suite now (also breaking centos). A fix is planned so I'm accepting broken tests for now rather than implementing a workaround only to revert it soon after.

ares commented:
Thanks everyone who put the effort on this, merged just in time :-)

metzgerd commented:
ansible-runner installation failed for ubuntu 18.04 because it is on this path not available for bionic.

Err:14 https://releases.ansible.com/ansible-runner/deb bionic Release
404 Not Found [IP: 104.26.1.250 443]

https://community.theforeman.org/t/foreman-ansible-plugin-installation-failed-with-unmet-dependencies/17799

ekohl commented:
@metzgerd please open a new issue for it. Comments on merged PRs are usually missed.
